pyvrp/tests/test_Population.py

```diff
@@ -189,56 +185,115 @@ def test_select_returns_same_parents_if_no_other_option():
+    elite_individuals = best_individuals[:nb_elite]
 
+    # Add a solution that is certainly not feasible, thus causing a purge.
+    single_route = [client for client in range(1, data.num_clients + 1)]
+    pop.add(Individual(data, pm, [single_route]))
 
-def test_restart_generates_min_pop_size_new_individuals():
-    """
-    Tests if restarting the population will generate ``min_pop_size`` new
-    individuals.
-    """
-    data = read("data/OkSmall.txt")
-    pm = PenaltyManager(data.vehicle_capacity)
-    rng = XorShift128(seed=12)
+    # After the purge, there should remain min_pop_size infeasible solutions.
+    assert_equal(pop.num_infeasible(), params.min_pop_size)
+
+    # In the infeasible subpopulation, nb_elite solutions from before the purge
+    # should also still be present. We test that by selecting the nb_elite best
+    # individuals from before the purge. We test by id/memory location because
+    # these individuals should still be present, unmodified.
+    new_individuals = [id(individual) for individual in pop]
+    for elite_individual in elite_individuals:
+        assert_(id(elite_individual) in new_individuals)
+
+
+def test_binary_tournament_ranks_by_fitness():
+    data = read("data/RC208.txt", "solomon", "dimacs")
+    pm = PenaltyManager(data.num_vehicles)
+    params = PopulationParams()
+    rng = XorShift128(seed=42)
 
-    params = PopulationParams(min_pop_size=2)
     pop = Population(data, pm, rng, broken_pairs_distance, params)
 
-    old_feas = {id(indiv) for indiv, *_ in pop.feasible_subpopulation}
-    old_infeas = {id(indiv) for indiv, *_ in pop.infeasible_subpopulation}
+    for _ in range(50):
+        pop.add(Individual(data, pm, rng))
 
-    pop.restart()
+    assert_equal(pop.num_feasible(), 0)
 
-    new_feas = {id(indiv) for indiv, *_ in pop.feasible_subpopulation}
-    new_infeas = {id(indiv) for indiv, *_ in pop.infeasible_subpopulation}
+    # Since this test requires the fitness values of the individuals, we have
+    # to access the underlying infeasible subpopulation directly.
+    infeas = [item for item in pop._infeas]
+    infeas = sorted(infeas, key=lambda item: item.fitness)
+    infeas = {item.individual: idx for idx, item in enumerate(infeas)}
+    infeas_count = np.zeros(len(infeas))
 
-    assert_equal(len(pop), 2)
-    assert_(new_feas != old_feas)
-    assert_(new_infeas != old_infeas)
+    for _ in range(10_000):
+        indiv = pop.get_binary_tournament()
+        infeas_count[infeas[indiv]] += 1
+
+    # Now we compare the observed ranking from the binary tournament selection
+    # against what we would expect from the actual fitness ranking. We compute
+    # the percentage of times we're incorrect, and test that that number is not
+    # too high.
+    actual_rank = np.argsort(-infeas_count)  # higher is better
+    expected_rank = np.arange(len(infeas))
+    pct_off = np.abs((actual_rank - expected_rank) / len(infeas)).mean()
+
+    assert_(pct_off < 0.05)
```

Comment:
Would it be possible to bring back the biased fitness computation test as well (I know it tests internals but so do some other tests here). https://github.com/N-Wouda/PyVRP/blob/population-generics/pyvrp/tests/test_Population.py#L345

While less convenient, we could look up the index for a random individual to define a custom diversity function so it should be possible. I am curious if that test would currently pass given that I think we miss an `updateFitness` call in `purge` (but I may have missed something here).

Reply:
I'll try to do this in a new `test_SubPopulation` file, but when I tried to do that earlier today it turned into a right mess.

Reply:
I've tried doing this, but my attempts so far have boiled down to basically testing whether the fitness values are computed exactly the way they're computed. That's a pretty poor test. I've so far added a few tests that checks the average distance of the closest few individuals, which is an important component in those calculations.

Reply:
I think it's not a poor test given that the implementation computes it differently (e.g. by caching closest indivs). Checking that this is consistent with a simple but inefficient implementation of the definition makes sense to me. This argument was stronger with the numpy implementation but I think it is still useful to have a good test on fitness values. I do think the fact that we can't test it easily without making a mess suggests things are too much entangled but given that we have other things to do as well I suggest we leave it for now.